File: Programmers/Lv2/skill_tree.cpp
#include <string>
#include <vector>
#include <map>

using namespace std;

int solution(string skill, vector<string> skill_trees) {
    int answer = 0;
    map<char,int> skill_order;
    int cnt=1;
    
    for(int i=0; i<skill.length(); i++)
       skill_order.insert(pair<char,int>(skill[i],i+1));


    for(auto str : skill_trees){
        answer++;
        cnt=1;
        for(int i=0; i<str.length(); i++){
            if(skill_order[str[i]] == cnt) cnt++;
            else if(skill_order[str[i]] > cnt) {
                answer--;
                break;
            }
        }
    }
    
    return answer;
}

File: java/src/programmers/lv2/skillTree.java
package programmers.lv2;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class skillTree {
    public int solution(String skill, String[] skill_trees) {
        int answer = 0;

        for(String skill_tree : skill_trees){
            if(check_skillTree(skill, skill_tree)) answer++;
        }

        return answer;
    }
    public boolean check_skillTree(String skill, String skill_tree){
        int idx = 0;
        for(char s : skill_tree.toCharArray()){
            if( skill.indexOf(s) == -1 ) continue;
            if( skill.charAt(idx) == s ) idx++;
            else return false;
        }
        return true;
    }

    @Test
    void test(){
        assertEquals(2, solution("CBD",new String[]{"BACDE", "CBADF", "AECB", "BDA"}));
        assertEquals(1, solution("BDA",new String[]{"BDA"}));
    }
}
